### ISSUE TITLE
Questions/Doubts Regarding TCAV

### ISSUE THREAD

**Issue by jssambee** [tensorflow/tcav#38]
Hi,
I have been testing the TCAV for a bit now and need some clarity on the following doubts.

1) How many random concepts are really needed to get a trustworthy/statistically significant result? Since running with smaller number of random runs(~50-100) does give statistically significant results but does not give consistent results across multiple runs and running it for 500 random runs is very much computationally expensive, and given that different concepts needs to be tested, makes it really hard.
2)  Why do we need to compute duplicate CAVs between random500_0 and random500_1 and vice-versa? isn't it the same? 

I know similar issues have been reported but probably I haven't understood the clear logic here.

**Comment by BeenKim** [comment]
Hello Jssmbee,

1. The goal of statistical testing is to have *some* safe guard against spurious results, but guaranteeing anything here is an open research question. That means, yes, the stat testing isn't perfect, but it's at least we can do. The cause of this uncertainty stems from the fact that we don't yet to have full understanding about high dimensional representations, although a big part of community is working on making progress. I do think we must do this testing to at least guard ourselves against spurious results (to some extend). I'm always open to new, less expensive ways to test this - let us know if you have ideas. 

Computational issues: The most expensive part of the computation from my experience is getting activations of concept and random pictures, which is completely parallizable. Once you get the activations, each random runs are also completely paralleliz-able - each random run only regards concept pictures and a set of random pictures. Each random run also only requires getting derivative and training a linear classifier, both of which are not computationally expensive (relatively speaking, we are not inverting any giant matrices..).

2. I'm not sure what you mean. we never compute a CAV with same sets of images. To compute random CAVs (that will be used for stat testing), you need two sets (positive set, negative set). Those are: (random500_0, random500_1), (random500_0, random500_2), ...(random500_0, random500_20). These aren't duplicates of each other, since each sets of random folders are different. Please let me know if you are asking different question.

**Comment by jssambee** [comment]
Hi @BeenKim ,
Thanks for the response. 
1) I agree on what you said. We have also been looking into it and believe that the gradients for a single set, e.g. random500_0 for a single bottleneck and a single target has the same gradient across tests, so instead of computing gradients every time for random0 against all random sets, we can only compute it once and cache it for rest of the tests. Let me know if i am missing anything here.

2) Apologies for the confusion. Well what I meant was that the TCAV score for random500_1 vs random500_0 should be the same as 1 - (random500_0 vs random500_1). So don't really understand why we need to do both the computations instead of just doing one?

**Comment by BeenKim** [comment]
1. I think you got that right! The current code isn't designed to do that (rather, we just parallelize for each (random 0, random i) tuple). In my experience, getting gradient is pretty cheap, but getting activations of target images and concept images is more expensive (probably some function of TF). So I recommend for your activation_generation function to first run over all images in parallel (both target and concept images) to the network, then cache the activations before running TCAV. I find that most effective in accelerating. Another tip you may already know: don't use pickle since they are super expensive to load. Use np array to save (np.save) - much cheaper. 

2. It's true that TCAV score for  (random500_1, random500_0)  is same as 1 - (random500_0, random500_1). However, we don't calculate that twice. To calculate random CAV TCAV scores, we do this by fixing positive set (random500_1) and changing negative sets. So: (random500_1, random500_0), (random500_1, random500_2), (random500_1, random500_3), (random500_1, random500_4)....(random500_1, random500_100). None of these are the same CAVs because negative sets are different. I might be still  missing something - please let me know!

**Comment by jssambee** [comment]
Thank you for the clarification and Tips @BeenKim !
